server/src/main/java/org/opensearch/common/lucene/index/OpenSearchDirectoryReader.java

```diff
@@ -68,7 +72,61 @@ public ShardId shardId() {
     @Override
     public CacheHelper getReaderCacheHelper() {
         // safe to delegate since this reader does not alter the index
-        return in.getReaderCacheHelper();
+        return this.delegatingCacheHelper;
+    }
+
+    public DelegatingCacheHelper getDelegatingCacheHelper() {
+        return this.delegatingCacheHelper;
+    }
+
+    /**
+     * Wraps existing IndexReader cache helper which internally provides a way to wrap CacheKey.
+     * @opensearch.internal
+     */
+    public class DelegatingCacheHelper implements CacheHelper {
+        CacheHelper cacheHelper;
+        DelegatingCacheKey serializableCacheKey;
+
+        DelegatingCacheHelper(CacheHelper cacheHelper) {
+            this.cacheHelper = cacheHelper;
+            this.serializableCacheKey = new DelegatingCacheKey(cacheHelper.getKey());
+        }
+
+        @Override
+        public CacheKey getKey() {
+            return this.cacheHelper.getKey();
+        }
+
+        public DelegatingCacheKey getDelegatingCacheKey() {
+            return this.serializableCacheKey;
+        }
+
+        @Override
+        public void addClosedListener(ClosedListener listener) {
+            this.cacheHelper.addClosedListener(listener);
+        }
+    }
+
+    /**
+     *  Wraps internal IndexReader.CacheKey and attaches a uniqueId to it which can be eventually be used instead of
+     *  object itself for serialization purposes.
+     */
+    public class DelegatingCacheKey {
+        CacheKey cacheKey;
+        private final UUID uniqueId;
+
+        DelegatingCacheKey(CacheKey cacheKey) {
+            this.cacheKey = cacheKey;
+            this.uniqueId = UUID.randomUUID();
+        }
+
+        public CacheKey getCacheKey() {
+            return this.cacheKey;
+        }
+
+        public UUID getId() {
+            return uniqueId;
+        }
```

Comment:
Given every cache helper object is associate with single cache key, combining into single class makes it more readable IMO. Ideally, DelegatingCacheKey should have been extension of CacheKey, but that class is final within Lucene.

```
    public class DelegatingCacheHelper implements CacheHelper {
        private final CacheHelper cacheHelper;
        private final UUID uniqueId;;

        DelegatingCacheHelper(CacheHelper cacheHelper) {
            this.cacheHelper = cacheHelper;
            this.uniqueId = UUID.randomUUID();
        }

        @Override
        public CacheKey getKey() {
            return this.cacheHelper.getKey();
        }

        public UUID getId() {
            return uniqueId;
        }

        @Override
        public void addClosedListener(ClosedListener listener) {
            this.cacheHelper.addClosedListener(listener);
        }
    }
```

Reply:
I don't have a strong opinion on this. What you suggested makes it only less verbose IMO, don't think readability is an issue. But from logical point of view it makes more sense to me to have uniqueId associated with CacheKey itself though there is 1:1 mapping between helper:key.

Reply:
While I don't feel strongly about it, but I don't see any place where `DelegatingCacheKey` is being used as is. It is always like `getKey().getId()` or `getKey().getCacheKey()`. So, I am questioning if the caller even needs to know about the existence of DelegatingCacheKey?

Reply:
It is not used right now, but will be used later on like in QueryCache which takes CacheKey as a key directly. Here we will be using this as it is. So I think we should keep it.

Reply:
Will the QueryCache need DelegatingCacheKey? If not, can we provide CacheKey using getKey() method.

Reply:
Yeah QueryCache will need DelegatingCacheKey.

Reply:
In the interim, we can expose `getId` method from the `DelegatingCacheHelper` and keep `getDelegatingCacheKey` private for now? You can make it public when you start using it. `getId` can be thought of as additional convenience method